torch/csrc/distributed/c10d/NanCheck.cu

```diff
@@ -11,13 +11,134 @@ namespace c10d {
 
 // CUDA kernel to check if data has NAN, device side assert
 // is raised if NAN is found
+
+// Using ulong2 as a "byte pack", with 16 bytes, for efficient data load
+typedef ulong2 BytePack;
+
+// AMD HIP doesn't define `__trap()`, using `assert` instead
+#ifdef USE_ROCM
+#define __trap() assert(0)
+#endif
+
+//// Start of templated functions for checking NaNs inside a BytePack
+
+// (i) General implementation (aka fallback)
+// We use a for loop to iterate over the elements in a BytePack.
+// EltPerPack would be greater than 8 if falling in this case.
+
+template <typename T, int EltPerPack>
+struct CheckBytePack {
+  static __device__ __forceinline__ void check(BytePack* tmp) {
+    T* data = (T*)tmp;
+    #pragma unroll 8
+    for (int i = 0; i < EltPerPack; i++) {
+      if (isnan(data[i])) __trap();
+    }
+  }
+};
+
+// (ii) Template Specialization for 8-byte data types, e.g. double
+// EltPerPack = 16 / 8 = 2
+
+template <typename T>
+struct CheckBytePack<T, /*EltPerPack*/2> {
+  static __device__ __forceinline__ void check(BytePack* tmp) {
+    T* data = (T*)tmp;
+    if (isnan(data[0]) || isnan(data[1])) __trap();
+  }
+};
+
+// (iii) Template specialization for 4-byte data types, e.g. float32
+// EltPerPack = 16 / 4 = 4
+
+template <typename T>
+struct CheckBytePack<T, /*EltPerPack*/4> {
+  static __device__ __forceinline__ void check(BytePack* tmp) {
+    T* data = (T*)tmp;
+    if (isnan(data[0]) || isnan(data[1]) || isnan(data[2]) || isnan(data[3])) __trap();
+  }
+};
+
+// (iv) Template specialization for 2-byte data types, e.g. float16, bfloat16, half.
+// EltPerPack = 16 / 2 = 8
+
+template <typename T>
+struct CheckBytePack<T, /*EltPerPack*/8> {
+  static __device__ __forceinline__ void check(BytePack* tmp) {
+    T* data = (T*)tmp;
+    if (isnan(data[0]) || isnan(data[1]) || isnan(data[2]) || isnan(data[3]) ||
+        isnan(data[4]) || isnan(data[5]) || isnan(data[6]) || isnan(data[7])) {
+          __trap();
+    }
+  }
+};
+
+//// End of templated functions for checking NaNs inside a BytePack
+
+
+// Fast-path check routine:
+// each thread will load and check 8 BytePacks in this routine
+
+// Create a tmp buffer of size 8, also unroll for loop by 8
+#define UNROLL 8
+
+template <typename T>
+__device__ __forceinline__ void checkChunk(BytePack* ptr) {
+  BytePack tmp[UNROLL];
+  int nWorkers = blockDim.x * gridDim.x;
+  // First load values from global memory into tmp buffer
+  #pragma unroll 8
+  for (int j = 0; j < UNROLL; j++) {
```

Comment:
hm, does checkChunk get called with a different ptr offset for each thread?

Reply:
Below at line 134, `checkChunk ` is called like this:
`checkChunk<T>(ptr + offset);`
`offset` accounts for different offsets for different threads.